sim/scenarios/corrupt-rate/corrupt-rate-error-model.cc

```diff
@@ -0,0 +1,63 @@
+#include <cstdint>
+
+#include "corrupt-rate-error-model.h"
+#include "ns3/header.h"
+#include "ns3/packet.h"
+#include "ns3/ppp-header.h"
+#include "ns3/ipv4-header.h"
+#include "ns3/udp-header.h"
+
+using namespace std;
+
+NS_OBJECT_ENSURE_REGISTERED(CorruptRateErrorModel);
+
+TypeId CorruptRateErrorModel::GetTypeId(void) {
+    static TypeId tid = TypeId("CorruptRateErrorModel")
+        .SetParent<ErrorModel>()
+        .AddConstructor<CorruptRateErrorModel>()
+        ;
+    return tid;
+}
+ 
+CorruptRateErrorModel::CorruptRateErrorModel()
+    : rate(0), distr(0, 99) {
+    std::random_device rd;
+    rng = new std::mt19937(rd());
+}
+
+void CorruptRateErrorModel::DoReset(void) { }
+
+bool CorruptRateErrorModel::DoCorrupt(Ptr<Packet> p) {
+    if (distr(*rng) >= rate) return false;
+
+    const uint32_t p_len = p->GetSize();
+    uint8_t *buffer= new uint8_t[p_len];
+    p->CopyData(buffer, p->GetSize());
+    PppHeader ppp_hdr = PppHeader();
+    uint32_t ppp_hdr_len = p->RemoveHeader(ppp_hdr);
+    Ipv4Header ip_hdr = Ipv4Header();
+    uint32_t ip_hdr_len = p->RemoveHeader(ip_hdr);
+    UdpHeader udp_hdr = UdpHeader();
+    uint32_t udp_hdr_len = p->RemoveHeader(udp_hdr);
+    const uint32_t hdr_len = ppp_hdr_len + ip_hdr_len + udp_hdr_len;
+    std::uniform_int_distribution<> d(hdr_len, hdr_len + p->GetSize() - 1);
+    auto pos = d(*rng);
+    buffer[pos] ^= 0xff;
+    cout << "Corrupting byte " << pos - hdr_len << " of the UDP payload (total: " << p->GetSize() << ")" << endl;
+    uint32_t payload_len = p_len - hdr_len;
+    Packet new_p = Packet(&buffer[hdr_len], payload_len);
```

Comment:
I'm not sure what the rest of this function is doing. Can you add a comment here?

Reply:
When it comes to header processing, ns3 only allows you to do one of two things:
1. peek the first header of a packet.
2. remove a header from a packet.

Peeking is fine if you're just interested in the outer-most layer, but if you need to look at multiple layers, it seems like there's no other way than removing the headers. So that's what I'm doing in the lines before this.
Having removed all the headers and flipped one byte, I now have to reassemble the packet by adding UDP, IP and PPP headers (and making sure that the respective checksums are recalculated, otherwise the packet will get dropped).

I'll add a comment.